File: hphp/hack/src/decl/decl_enum.ml
(*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

open Hh_prelude
open Aast
open Typing_defs
module SN = Naming_special_names

type t = {
  base: Typing_defs.decl_ty;
      (** Underlying type of the enum, e.g. int or string.
          For subclasses of Enum, this is the type parameter of Enum.
          For enum classes, this is HH\MemberOf<E, I>. *)
  type_: Typing_defs.decl_ty;
      (** Type containing the enum name.
          For subclasses of Enum, this is also the type parameter of Enum. *)
  constraint_: Typing_defs.decl_ty option;
      (** Reflects what's after the [as] keyword in the enum definition. *)
  interface: Typing_defs.decl_ty option;
      (** For enum classes, this is the raw interface I, as provided by the user. *)
}

(** Figures out if a class needs to be treated like an enum. *)
let enum_kind name ~is_enum_class enum inner_ty ~get_ancestor =
  match enum with
  | None ->
    (match get_ancestor SN.FB.cEnum with
    | Some enum -> begin
      match get_node enum with
      | Tapply ((_, enum), [ty_exp]) when String.equal enum SN.FB.cEnum ->
        Some
          {
            base = ty_exp;
            type_ = ty_exp;
            constraint_ = None;
            interface = None;
          }
      | Tapply ((_, enum_class), _) when String.equal enum_class SN.FB.cEnum ->
        let ty_exp =
          (* The fallback if the class does not declare TInner (i.e. it is
           * abstract) is to use this::TInner
           *)
          match inner_ty with
          | None ->
            let this = Typing_defs_core.mk (get_reason enum, Tthis) in
            Typing_defs_core.mk
              (get_reason enum, Taccess (this, (get_pos enum, SN.FB.tInner)))
          | Some ty -> ty
        in
        Some
          {
            base = ty_exp;
            type_ = ty_exp;
            constraint_ = None;
            interface = None;
          }
      | _ -> None
    end
    | _ -> None)
  | Some enum ->
    let reason = get_reason enum.te_base in
    let pos = Reason.to_pos reason in
    let enum_type = Typing_defs.mk (reason, Tapply (name, [])) in
    let (te_base, te_interface) =
      if is_enum_class then
        let te_interface = enum.te_base in
        let te_base =
          Tapply ((pos, SN.Classes.cMemberOf), [enum_type; enum.te_base])
        in
        (* TODO(T77095784) make a new reason ! *)
        let te_base = mk (reason, te_base) in
        (te_base, Some te_interface)
      else
        (enum.te_base, None)
    in
    Some
      {
        base = te_base;
        type_ = Typing_defs.mk (get_reason enum.te_base, Tapply (name, []));
        constraint_ = enum.te_constraint;
        interface = te_interface;
      }

(** If a class is an Enum, we give all of the constants in the class the type
    of the Enum. We don't do this for Enum<mixed> and Enum<arraykey>, since
    that could *lose* type information. *)
let rewrite_class name ~is_enum_class enum inner_ty ~get_ancestor consts =
  match enum_kind name ~is_enum_class enum inner_ty ~get_ancestor with
  | None -> consts
  | Some { base = _; type_ = ty; constraint_ = _; interface = te_interface } ->
    let te_enum_class = Option.is_some te_interface in
    (match get_node ty with
    | Tmixed
    | Tprim Tarraykey ->
      consts
    | _ ->
      (* A special constant called "class" gets added, and we don't
       * want to rewrite its type.
       * Also for enum class, the type is set in the lowerer.
       *)
      SMap.mapi
        (fun k c ->
          if te_enum_class || String.equal k SN.Members.mClass then
            c
          else
            { c with cc_type = ty })
        consts)

File: hphp/hack/src/hackrs/folded_decl_provider/fold/decl_enum.rs
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.
use pos::ClassConstNameIndexMap;
use pos::Positioned;
use pos::TypeNameIndexMap;
use special_names as sn;
use ty::decl::folded::ClassConst;
use ty::decl::Prim;
use ty::decl::Ty;
use ty::decl::Ty_;
use ty::reason::Reason;

use super::DeclFolder;

struct EnumKind<R: Reason> {
    /// Underlying type of the enum, e.g. int or string. For subclasses of
    /// `Enum`, this is the type parameter of the Enum. For enum classes, this
    /// is `HH\MemberOf<E, I>`.
    // NB(jakebailey, 2022-03-11): `base` is copied from OCaml but not used here.
    // base: Ty<R>,

    /// Type containing the enum name.
    /// For subclasses of Enum, this is also the type parameter of Enum.
    ty: Ty<R>,

    /// Reflects what's after the `as` keyword in the enum definition.
    // NB(jakebailey, 2022-03-11): `constraint` is copied from OCaml but not used here.
    // constraint: Option<Ty<R>>,

    /// For enum classes, this is the raw interface I, as provided by the user.
    interface: Option<Ty<R>>,
}

impl<'a, R: Reason> DeclFolder<'a, R> {
    /// Figures out if `self.child` needs to be treated like an enum.
    fn enum_kind(
        &self,
        inner_ty: Option<&Ty<R>>,
        ancestors: &TypeNameIndexMap<Ty<R>>,
    ) -> Option<EnumKind<R>> {
        let is_enum_class = matches!(self.child.kind, ty::decl::ty::ClassishKind::CenumClass(..));
        match &self.child.enum_type {
            None => {
                let enum_ty = match ancestors.get(&*sn::fb::cEnum) {
                    None => return None,
                    Some(ty) => ty,
                };
                match enum_ty.unwrap_class_type() {
                    (_, name, [ty_exp]) if name.id() == *sn::fb::cEnum => Some(EnumKind {
                        // base: ty_exp.clone(),
                        ty: ty_exp.clone(),
                        // constraint: None,
                        interface: None,
                    }),
                    (_, name, _) if name.id() == *sn::fb::cEnum => {
                        // The fallback if the class does not declare `TInner` (i.e.
                        // it is abstract) is to use `this::TInner`
                        let r = || enum_ty.reason().clone();
                        let ty_exp = match inner_ty {
                            Some(ty) => ty.clone(),
                            None => Ty::access(
                                r(),
                                ty::decl::TaccessType {
                                    ty: Ty::this(r()),
                                    type_const: Positioned::new(
                                        enum_ty.pos().clone(),
                                        *sn::fb::tInner,
                                    ),
                                },
                            ),
                        };
                        Some(EnumKind {
                            // base: ty_exp.clone(),
                            ty: ty_exp,
                            // constraint: None,
                            interface: None,
                        })
                    }
                    _ => None,
                }
            }
            Some(enum_type) => {
                let reason = enum_type.base.reason();
                let pos = reason.pos();
                let enum_ty = Ty::apply(reason.clone(), self.child.name.clone(), [].into());
                let (te_base, te_interface) = if is_enum_class {
                    let te_interface = enum_type.base.clone();
                    // TODO(T77095784) make a new reason !
                    let te_base = Ty::apply(
                        reason.clone(),
                        Positioned::new(pos.clone(), *sn::classes::cMemberOf),
                        [enum_ty, enum_type.base.clone()].into(),
                    );
                    (te_base, Some(te_interface))
                } else {
                    (enum_type.base.clone(), None)
                };
                Some(EnumKind {
                    ty: Ty::apply(te_base.reason().clone(), self.child.name.clone(), [].into()),
                    // base: te_base,
                    // constraint: enum_type.constraint.clone(),
                    interface: te_interface,
                })
            }
        }
    }

    /// If `self.child` is an Enum, we give all of the constants in the class
    /// the type of the Enum. We don't do this for `Enum<mixed>` and
    /// `Enum<arraykey>`, since that could *lose* type information.
    pub fn rewrite_class_consts_for_enum(
        &self,
        inner_ty: Option<&Ty<R>>,
        ancestors: &TypeNameIndexMap<Ty<R>>,
        consts: &mut ClassConstNameIndexMap<ClassConst<R>>,
    ) {
        let EnumKind {
            // base: _,
            ty,
            // constraint: _,
            interface,
        } = match self.enum_kind(inner_ty, ancestors) {
            None => return,
            Some(kind) => kind,
        };

        // Don't rewrite enum classes.
        if interface.is_some() {
            return;
        }

        // Don't rewrite `Enum<mixed>` or `Enum<arraykey>`.
        if matches!(ty.node_ref(), Ty_::Tmixed | Ty_::Tprim(Prim::Tarraykey)) {
            return;
        }

        // A special constant called "class" gets added, and we don't
        // want to rewrite its type.
        // Also for enum class, the type is set in the lowerer.
        for (&name, c) in consts.iter_mut() {
            if name != *sn::members::mClass {
                c.ty = ty.clone();
            }
        }
    }
}
